### PR TITLE
Stop suggesting already assigned approvers

### PR DESCRIPTION
**Pull request by matthyx**
Relates to #16793
This is a complete departure from the existing logic, it probably needs confirmation from @cjwagner before merging...
/cc @BenTheElder 
/cc @cjwagner

### CODE REVIEW

**Comment by cjwagner** [comment]
/hold
I don't think this is what we want. This is the concern:
> Seems a bit silly to suggest assigning someone who is already assigned, if nothing else? Perhaps we can just not suggest then?

I think the core of the concern is that it is confusing to be asked to assign someone who is already assigned, not that we should avoid suggesting getting approval from assignees.
If a user with sufficient approval permissions is already assigned to a PR I think we want to continue suggesting getting approval from them rather than suggesting another person who isn't assigned to the PR. BUT we should recognize when the user is already assigned and suggest prompting them for approval rather than `/assign`ing them again since that does nothing and is confusing/silly.

**Comment by matthyx** [comment]
> we should recognize when the user is already assigned and suggest prompting them for approval rather than `/assign`ing them again since that does nothing and is confusing/silly.

ok, so I keep the current behavior of `GetCCs` but in the comment template itself I treat this information differently if the suggested person is already assigned

**Comment by matthyx** [comment]
Like that @cjwagner ?

File: prow/plugins/approve/approve_test.go
Comment (chaodaiG) [inline prow/plugins/approve/approve_test.go:940]: Is there a test case that cjwagner is not assigned on this PR yet, and the comment still suggests `/assign`?

File: prow/plugins/approve/approvers/owners.go
Comment (chaodaiG) [inline prow/plugins/approve/approvers/owners.go:559]: Should we have unit test for these 2 new functions?

**Review by chaodaiG — COMMENTED** [review]
Overall looks good to me, feel like there are some unit tests missing

File: prow/plugins/approve/approvers/owners.go
Comment (cjwagner) [inline prow/plugins/approve/approvers/owners.go:648]: If `GetCCs()` returns multiple users and some are assignees while others are not, we'll output both of these sections which seems wrong to me. Please add a unit test for this case too.

Comment (matthyx) [reply prow/plugins/approve/approvers/owners.go:648]: It was on purpose... What do you suggest then?

Comment (cjwagner) [reply prow/plugins/approve/approvers/owners.go:648]: Correction: I meant confusing not wrong. I think we **do** want to recommend both assigning the SuggestedCCs and requesting approval from the AssignedCCs, but I think it is confusing to have two lines with different instructions for "To complete the pull request process...".

To me it would be clearer to have one set of instructions when we have both suggested and assigned CCs. Something like this:
> To complete the pull request process, please ask for approval from cjwagner and additionally assign mattyx after the PR has been reviewed.
You can assign the PR to them by writing `/assign matthyx` in a comment when ready

Comment (matthyx) [reply prow/plugins/approve/approvers/owners.go:652]: done this part for review, although I need to add some test cases suggested by @chaodaiG

File: prow/plugins/approve/approve_test.go
Comment (matthyx) [reply prow/plugins/approve/approve_test.go]: like prow/plugins/approve/approvers/approvers_test.go:866 ?

Comment (chaodaiG) [reply prow/plugins/approve/approve_test.go]: Yeah that would work. More specifically:
-Without- the line containing `/approve cancel`
-With- the line containing `"someone else", "github.ReviewStateApproved` or `"someone else", "/approve"

Also you can test when there is only self-approval

File: prow/plugins/approve/approvers/owners.go
Comment (MushuEE) [reply prow/plugins/approve/approvers/owners.go:559]: nit: quick doc string for public functions

**Comment by matthyx** [comment]
Sorry for the delay, PTAL!

File: prow/plugins/approve/approve_test.go
Comment (chaodaiG) [inline prow/plugins/approve/approve_test.go:1106]: it took me a while to figure out the logic of this unit test:
- cjwagner is the author of all test cases in this test
- this test case has `selfApprove: false`, so cjwagner didn't approve their own test
- then the PR suggests ask for approval from cjwagner

Logically speaking this is all correct, and this is perhaps the best we can do for this scenario in prod. Might not be in this PR, but maybe we should make the PR author configurable to cover more scenarios?

**Review by chaodaiG — APPROVED** [review]
LGTM, left a comment

/hold
In case you want @cjwagner to take another look

**Review by cjwagner — APPROVED** [review]

**Pull request merged**
